internal/orchestrator/orchestrator.go

```diff
@@ -403,204 +424,461 @@ func (or *orchestrator) initDataExchange(ctx context.Context) (err error) {
+		return nil
+	}
+
+	// If there still is no tokens config, check the deprecated structure for config
+	tokensConfigArraySize = deprecatedTokensConfig.ArraySize()
+	if tokensConfigArraySize > 0 {
+		log.L(ctx).Warnf("Your tokens config uses a deprecated configuration structure - the tokens configuration has been moved under the 'plugins' section")
+	}
+
+	for i := 0; i < tokensConfigArraySize; i++ {
+		prefix := deprecatedTokensConfig.ArrayEntry(i)
+		name := prefix.GetString(tokens.TokensConfigName)
+		pluginName := prefix.GetString(tokens.TokensConfigPlugin)
+		if name == "" {
+			return i18n.NewError(ctx, coremsgs.MsgMissingTokensPluginConfig)
+		}
+		if err = core.ValidateFFNameField(ctx, name, "name"); err != nil {
+			return err
+		}
+		if pluginName == "" {
+			// Migration path for old config key
+			// TODO: eventually make this fatal
+			pluginName = prefix.GetString(tokens.TokensConfigConnector)
 			if pluginName == "" {
-				// Migration path for old config key
-				// TODO: eventually make this fatal
-				pluginName = config.GetString(tokens.TokensConfigConnector)
-				if pluginName == "" {
-					return i18n.NewError(ctx, coremsgs.MsgMissingTokensPluginConfig)
-				}
-				log.L(ctx).Warnf("Your tokens config uses the deprecated 'connector' key - please change to 'plugin' instead")
-			}
-			if pluginName == "https" {
-				// Migration path for old plugin name
-				// TODO: eventually make this fatal
-				log.L(ctx).Warnf("Your tokens config uses the old plugin name 'https' - this plugin has been renamed to 'fftokens'")
-				pluginName = "fftokens"
+				return i18n.NewError(ctx, coremsgs.MsgMissingTokensPluginConfig)
 			}
+			log.L(ctx).Warnf("Your tokens config uses the deprecated 'connector' key - please change to 'plugin' instead")
+		}
+		if pluginName == "https" {
+			// Migration path for old plugin name
+			// TODO: eventually make this fatal
+			log.L(ctx).Warnf("Your tokens config uses the old plugin name 'https' - this plugin has been renamed to 'fftokens'")
+			pluginName = "fftokens"
+		}
 
-			log.L(ctx).Infof("Loading tokens plugin name=%s plugin=%s", name, pluginName)
-			plugin, err := tifactory.GetPlugin(ctx, pluginName)
-			if plugin != nil {
-				err = plugin.Init(ctx, name, config, &or.bc)
-			}
-			if err != nil {
-				return err
-			}
-			or.tokens[name] = plugin
+		log.L(ctx).Infof("Loading tokens plugin name=%s plugin=%s", name, pluginName)
+		plugin, err := tifactory.GetPlugin(ctx, pluginName)
+		if plugin != nil {
+			err = plugin.Init(ctx, name, prefix, &or.bc)
+		}
+		if err != nil {
+			return err
 		}
+		or.tokens[name] = plugin
 	}
-
 	return nil
 }
 
 func (or *orchestrator) initComponents(ctx context.Context) (err error) {
 
 	if or.data == nil {
-		or.data, err = data.NewDataManager(ctx, or.database, or.sharedstorage, or.dataexchange)
+		or.data, err = data.NewDataManager(ctx, or.databases["database_0"], or.sharedstorage["sharedstorage_0"], or.dataexchange["dataexchange_0"])
```

Comment:
We can't assume there is a database named `database_0` if we are using a well-formed new config file, right?

Wonder if you actually need to keep the singletons like `or.database` and initialize them to the first plugin you find when doing the array init.

Side note, but you could also go as far as to reject config with more than one plugin of a given type (except for tokens), since right now it's ignored. Fine either way (reject it or just ignore it).

Reply:
For now, I'll just continue to ignore additional plugin configs